### ISSUE TITLE
Case where a machine is up but unreachable

### ISSUE THREAD

**Issue by Matansegal** [krojew/cdrs-tokio#158]
We have a node which is physically up (the machine is on), but not accepting connections.
We're seeing connection time outs, but the retry policy nor the reconnection policy seem to trigger.

When we tested it, it goes through this line:
https://github.com/krojew/cdrs-tokio/blob/cb4253500ceb2bef46e71691bf63ab84dc29848d/cdrs-tokio/src/cluster/send_envelope.rs#L38

In such a case, shouldn't it move to the next node and mark this one as temporary down?

**Comment by krojew** [comment]
Are you sure the reconnection policy was not triggered? https://github.com/krojew/cdrs-tokio/blob/master/cdrs-tokio/src/cluster/tcp_connection_manager.rs#L58 should ask to jump to the next node if connection was not successful. What reconnection policy was used?

**Comment by Matansegal** [comment]
It doesn't seem like it get to this error even though we still get the error for timeout. It gets in to the loop.
We tried all reconnection policies available. All haven't handle this issue.

**Comment by krojew** [comment]
O, I'll try to replicate the issue. Can you share the exact error you're getting?

**Comment by Matansegal** [comment]
The error we got is `IO error: Connection timed out (os error 110)`
When we limit the connection pool config to 100ms (in order to be able to test it and not wait a whole minute for getting back from this error) we receive the error: ` Timeout: Timeout waiting for connection to: XX.XX.XXX.XXX:XXXX`

**Comment by Matansegal** [comment]
When I leave the service up for a while, it does get into this lines of error:
https://github.com/krojew/cdrs-tokio/blob/master/cdrs-tokio/src/cluster/tcp_connection_manager.rs#L58
and I print the problematic connection. However, the status of this None is still up instead of mark to be down or unknown.

**Comment by krojew** [comment]
After some testing, I don't think there's any error with handling the timeout itself (OS error 110). It can happen, for whatever reason, but then is caught by the `ConnectionManager`, which in turn asks the reconnection policy what to do. The only way for the error to be propagated up is for the policy to stop reconnecting.
The other type of timeout is related to connection timeout handled by the `ConnectionPool` - this one is user-defined and acts as a safety switch irrespective of reconnection policy.
Can you share your code or a minimal example which shows the error?

**Comment by krojew** [comment]
Ok, I'll dig some more into this issue. Thanks for the samples!

**Comment by Matansegal** [comment]
When I print the `addr` in: https://github.com/krojew/cdrs-tokio/blob/master/cdrs-tokio/src/cluster/tcp_connection_manager.rs#L56-L60
It print the dead address but it still try to reach it when I send requests.

**Comment by Matansegal** [comment]
It is similar solution to the one I suggested in term of not really solving the problem but having a temporary solution.
I get a response but since I have 6 nodes, on average every 6th request take much longer (depend on the time I set for timeout). We want to mark this node as down and not just skip it every time it reaches it.

**Comment by krojew** [comment]
That is the expected behavior after the latest changes. The next part is to implement the new node state, which should solve everything.

**Comment by Matansegal** [comment]
I found this in the cpp driver https://github.com/datastax/cpp-driver/blob/07f8adee80845d403ff6fc4fc8dcc6654a97cddd/src/control_connection.cpp

Didn't really get into it, but there is a refresh mechanism which I think should be useful .

**Comment by krojew** [comment]
> Didn't really get into it, but there is a refresh mechanism which I think should be useful .

This seems to be related only to topology changes in the control connection, which is not the case in our situation.

**Comment by krojew** [comment]
After some analysis, looks like the internal reconnection mechanism needs to change a bit. I will close your PR for now, since those changes will make it obsolete.

**Comment by Matansegal** [comment]
I got  `'attempt to calculate the remainder with a divisor of zero'`

for https://github.com/krojew/cdrs-tokio/blob/new-reconnection/cdrs-tokio/src/cluster/connection_pool.rs#L365

**Comment by Matansegal** [comment]
I made some tests.
We have 6 machines (nodes) for our database.
When just one node is down it work perfect. However, when I take another node down, it start marking all of the node as down and out of 1000 request I get response for less than 50. I tried with both ways I mentioned earlier here https://github.com/krojew/cdrs-tokio/issues/158#issuecomment-1503432570

**Comment by krojew** [comment]
That's extremely hard for me to test. Is there a possibility you can add some logging and gather what's going on or debug the internals?

**Comment by krojew** [comment]
The two dead nodes should be marked as down, but the rest should not. Can you add a breakpoint to `Node::mark_down()` and see what causes the remaining nodes to go down?

**Comment by krojew** [comment]
Ok, I found the issue - the heartbeat code is missing from this experimental branch...

**Comment by Matansegal** [comment]
Thank you! please let me know when you commit it.

**Comment by krojew** [comment]
Just merged the hearbeat code. Active connections should not go down anymore. You might need to adjust heartbeat interval in pool config if they still do.

**Comment by krojew** [comment]
@Matansegal have you had a chance to test the changes?

**Comment by Matansegal** [comment]
I just tried it and it doesn't seem to change the last outcomes. It still mark the live nodes as dead after several requests. I have mechanism to reduce level of consistency but it still does it until eventually I have no live nodes.  you can see the log I got below:

```
[2023-04-20T00:43:07.805 ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Up), }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Up), }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Up),  }
[2023-04-20T00:43:07.934 ] OUT
[2023-04-20T00:43:07.937 ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Up),  }
[2023-04-20T00:43:08.133 WARN ] Reducing level of consistency due cdrs_tokio::error::Error::Server=UnavailableError { cl: EachQuorum, required: 2, alive: 1 }
[2023-04-20T00:43:08.133 INFO ] Consistency level=Quorum
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Up),  }
[2023-04-20T00:43:08.233 WARN ] Reducing level of consistency due cdrs_tokio::error::Error::Server=UnavailableError { cl: Quorum, required: 2, alive: 1 }
[2023-04-20T00:43:08.233 INFO ] Consistency level=LocalQuorum
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Down),  }
[2023-04-20T00:43:08.233 ERROR ] Error accured while impl query for deal(s)=G2_MA0910 period(s)=201902; General error: No nodes available in query plan!
```

**Comment by krojew** [comment]
Did you use the latest version with the heartbeats? If so, maybe try lowering the hearbeat interval. The nodes are going down because you most likely have some kind of proxy dropping idle connections and the reconnection policy takes too long to get them back up. Using lower hearbeat and reconnection intervals should help.

**Comment by krojew** [comment]
I also pushed a change which is less strict in determining if a node is up or down, which might help in such scenarios.

**Comment by Matansegal** [comment]
I tried it with heartbeat of 1 sec then 300 ms, still got the errors.
The log below is with the DEBUG mode. it print debug line that I cannot put here.
IN is incomming request to the service. OUT is expected output.

```
[2023-04-20T14:13:51.895 INFO ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Up),  }
[2023-04-20T14:13:52.002 ERROR ] the Pool Level database does not contain a row with the deal=FR_RE6115 and period=202010
[2023-04-20T14:13:52.006 INFO ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Up),  }
[2023-04-20T14:13:52.108 WARN ] Reducing level of consistency due cdrs_tokio::error::Error::Server=UnavailableError { cl: EachQuorum, required: 2, alive: 1 }
[2023-04-20T14:13:52.109 INFO ] Consistency level=Quorum
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Up),  }
[2023-04-20T14:13:52.214 WARN ] Reducing level of consistency due cdrs_tokio::error::Error::Server=UnavailableError { cl: Quorum, required: 2, alive: 1 }
[2023-04-20T14:13:52.214 INFO ] Consistency level=LocalQuorum
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Up),  }
[2023-04-20T14:13:52.217 WARN ] Reducing level of consistency due cdrs_tokio::error::Error::Server=UnavailableError { cl: LocalQuorum, required: 2, alive: 1 }
[2023-04-20T14:13:52.217 ERROR ] Got to the lowest level of consistency (LocalQuorum) and still failing
[2023-04-20T14:13:52.220 ERROR ] Error accured while impl query; error: ErrorBody { message: "Cannot achieve consistency level LOCAL_QUORUM", ty: Unavailable(UnavailableError { cl: LocalQuorum, required: 2, alive: 1 }) }
[2023-04-20T14:13:52.329 INFO ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Down),  }
[2023-04-20T14:13:52.339 INFO ] OUT
[2023-04-20T14:13:52.342 INFO ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Down),  }
[2023-04-20T14:13:52.353 INFO ] OUT
[2023-04-20T14:13:52.380 INFO ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Down),  }
[2023-04-20T14:13:52.393 INFO ] OUT
[2023-04-20T14:13:52.396 INFO ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Down),  }
[2023-04-20T14:13:52.406 INFO ] OUT
[2023-04-20T14:13:52.409 INFO ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Down),  }
[2023-04-20T14:13:52.418 INFO ] OUT
[2023-04-20T14:13:52.421 INFO ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Up),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Down),  }
[2023-04-20T14:13:52.424 WARN ] Reducing level of consistency due cdrs_tokio::error::Error::Server=UnavailableError { cl: EachQuorum, required: 2, alive: 1 }
[2023-04-20T14:13:52.424 INFO ] Consistency level=Quorum
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Down),  }
[2023-04-20T14:13:52.427 ERROR ] Error accured while impl query; General error: No nodes available in query plan!
[2023-04-20T14:13:52.432 INFO ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Down),  }
[2023-04-20T14:13:52.434 ERROR ] Error accured while impl query; General error: No nodes available in query plan!
[2023-04-20T14:13:52.438 INFO ] IN
Node { broadcast_rpc_address: 11.11.111.111, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.112, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.113, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.114, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.115, state: Atomic(Down),  }
Node { broadcast_rpc_address: 11.11.111.116, state: Atomic(Down),  }
[2023-04-20T14:13:52.439 ERROR ] Error accured while impl query; General error: No nodes available in query plan!
```

**Comment by krojew** [comment]
There was one more option where the node could have been incorrectly marked as down and I just pushed a fix - can you test it? Also, please make sure you are using the new branch with a reconnection policy (the default one is ok) and there are debug logs - there should be a lot of them when the reconnection policy kicks in: https://github.com/krojew/cdrs-tokio/blob/new-reconnection/cdrs-tokio/src/cluster/connection_pool.rs

**Comment by krojew** [comment]
Are you absolutely sure you are using this branch and the debug logs are enabled? You should at least see hearbeat errors if a connection is down, which are absent from what you've pasted.

**Comment by Matansegal** [comment]
I am using this branch, we have the logging of the service. I am still trying to get the loggings form the driver.

**Comment by krojew** [comment]
Your logs got me thinking - the only way for a node to be marked as down, ignoring the reconnection policy, is via a server event. If you take a look at `cdrs_tokio::error::Error::Server=UnavailableError { cl: EachQuorum, required: 2, alive: 1 }` log, it means the cluster cannot find neighboring nodes. This suggest there might be something wrong with the cluster and it does indeed send a node down server event. Could that be the case?

**Comment by Matansegal** [comment]
I dont think so. Those servers are used by services written in python and cpp and do not have those errors.

**Comment by krojew** [comment]
I see. Well, those debug logs should show what's going on.

**Comment by krojew** [comment]
I'm testing this branch with my small test cluster and everything seems to work:

- no connections are dropped by themselves in time
- killing and restarting nodes bring the connections up per reconnection policy
- inserting a proxy in between and using it to block connections while keeping the node alive (which seems to be your case) also triggers reconnection policy and everything gets back up

I could really use those debug logs, since it seems you stumbled upon some weird edge case, which I cannot reproduce.

**Comment by krojew** [comment]
I found a possible edge case - the cluster might send node down event, while the node is in fact still up and there are active connections (which explains why the reconnection policy is not triggered). I pushed a change which will not mark such node as down, but it suggests something is happening with the cluster itself. Can you check?

**Comment by krojew** [comment]
All my tests show this branch is working. I hope your case if fixed too @Matansegal, but without your confirmation, I can only assume that. I will yank the current version and replace it with this one.

**Comment by krojew** [comment]
@Matansegal I've yanked the 8.0.0 due to this case and released 8.1.0. if the nodes are still being marked down, I really need those debug logs to see what the cause. At the moment I'm suspecting the node, to which the control connection is established, looses connection with other nodes and starts sending down events. In the meantime, something (a proxy?) is closing established connections from the driver if the heartbeat takes too long.
